### PR TITLE
[cmake] Add support for importing MODULE_LIBRARY targets and clean th…

### PR DESCRIPTION
**Pull request by gottesmm**
#### Resolved bug number: ([SR-](https://bugs.swift.org/browse/SR-))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

…ings up a little bit.

### CODE REVIEW

**Comment by gottesmm** [comment]
@swift-ci Please test

File: cmake/modules/SwiftSharedCMakeConfig.cmake [hunk 1]
```diff
@@ -42,13 +42,33 @@ endfunction()
...
 endfunction()
 
+function(get_imported_library_path outvar target target_type)
+  if ("${target_type}" STREQUAL "EXECUTABLE")
+    set(${outvar} "${LLVM_BINARY_OUTPUT_INTDIR}/${target}" PARENT_SCOPE)
+    return()
+  endif()
+
+  # LLVM compiles module libraries as
+  # ${MODULE_NAME}${CMAKE_SHARED_LIBRARY_SUFFIX} on both Linux and macOS.
```
Comment (jrose-apple) [inline cmake/modules/SwiftSharedCMakeConfig.cmake]: It looks like the correct variable is [CMAKE_SHARED_MODULE_SUFFIX](https://cmake.org/cmake/help/v3.0/variable/CMAKE_SHARED_MODULE_SUFFIX.html). Is there really no less ad hoc way to do this?

File: cmake/modules/SwiftSharedCMakeConfig.cmake [hunk 2]
```diff
@@ -63,27 +83,22 @@ function(fix_imported_target_locations_for_xcode targets)
...
+    # Set the target location properties.
+    if (NOT EXISTS "${PATH}")
+      message(WARNING "Target ${t} has an import location where no file "
+        "exists. Location: ${PATH}")
+    endif()
     set_target_properties(${t} PROPERTIES
-      IMPORTED_LOCATION_DEBUG "${LLVM_LIBRARY_OUTPUT_INTDIR}/${PREFIX}${t}.${CMAKE_${TARGET_TYPE}_SUFFIX}"
-      IMPORTED_LOCATION_RELEASE "${LLVM_LIBRARY_OUTPUT_INTDIR}/${PREFIX}${t}.${CMAKE_${TARGET_TYPE}_SUFFIX}")
+      IMPORTED_LOCATION_DEBUG "${PATH}"
+      IMPORTED_LOCATION_RELEASE "${PATH}")
```
Comment (jrose-apple) [inline cmake/modules/SwiftSharedCMakeConfig.cmake]: What about RELWITHDEBINFO? (I don't care so much about MINSIZEREL.)

Comment (gottesmm) [inline cmake/modules/SwiftSharedCMakeConfig.cmake]: I am fine adding that.

File: cmake/modules/SwiftSharedCMakeConfig.cmake (hunk 1 above)
Comment (gottesmm) [inline cmake/modules/SwiftSharedCMakeConfig.cmake]: Actually that is incorrect here. On APPLE,  CMAKE_SHARED_MODULE_SUFFIX is set to .so. I am not sure right now why that is broken in this way, but the warning will tell us if things change and we can no longer find it.

**Comment by gottesmm** [comment]
Jordan and I talked about this yesterday and came up with a different approach that is a little cleaner and allows me to remove the libclang hack.

**Comment by gottesmm** [comment]
@swift-ci Please test

**Comment by mattgallagher** [comment]
There's a remaining bug in this pull request... line 283 needs the same fix_imported_target_locations_for_xcode -> fix_imported_targets_for_xcode applied to line 190. After that, it appears to work fine for me.

**Comment by gottesmm** [comment]
I just checked locally on my machine. Everything passes.

**Comment by gottesmm** [comment]
@mattgallagher I know. I have that fixed locally. Thanks though!

**Comment by jrose-apple** [comment]
I like this a lot better. LGTM, thanks Michael!

**Comment by gottesmm** [comment]
I just updated the patch with my local changes that make everything work.

**Comment by gottesmm** [comment]
@jrose-apple I also like it better. The key thing that won me over was removing the libclang hack.

Isn't it amazing what two smart people with good intentions can conjure up? = ).

**Pull request merged**